src/typechecker.erl

```diff
@@ -4162,12 +4177,22 @@ type_check_function(Env, {function, _, Name, NArgs, Clauses}) ->
     case maps:find({Name, NArgs}, Env#env.fenv) of
         {ok, FunTy} ->
             NewEnv = Env#env{current_spec = FunTy},
-            FunTyNoPos = [ typelib:remove_pos(Ty) || Ty <- FunTy ],
-            check_clauses_fun(NewEnv, expect_fun_type(NewEnv, FunTyNoPos), Clauses);
+            FunTyNoPos = [ typelib:remove_pos(?assert_type(Ty, type())) || Ty <- FunTy ],
+            Arity = clause_arity(hd(Clauses)),
+            check_clauses_fun(NewEnv, expect_fun_type(NewEnv, FunTyNoPos, Arity), Clauses);
         error ->
             throw(internal_error(missing_type_spec, Name, NArgs))
     end.
 
+-spec clause_arity({clause, _, list(), _, _}) -> arity().
+clause_arity({clause, _, Args, _, _}) ->
+    arity(length(Args)).
+
+-spec arity(non_neg_integer()) -> arity().
+arity(I) ->
+    ?assert(I < 256, arity_overflow),
+    ?assert_type(I, arity()).
```

Comment:
Cool.

Is this what we ideally would want ?assert_type to do for all types in runtime?

Reply:
Well, a bit tricky question. If we use `mfa()` or `arity()` in any specs or types, and `erl_parse:abstract_type/0` does, then we run into the problem of convincing the type checker that any arbitrary integers we try to use as arities make sense, i.e. fall into 0..256 range. We could do the type assertion without the dynamic check, but it could be unsound. The way it's done here we have the assertion in place for testing, but we can still drop the dynamic check for release by defining `NOASSERT`.

From a different perspective, if we want to make the Gradualizer source code itself be an example of how to write as safe as possible code with Gradualizer, then I think this is what we would write to stay on the safe/conservative side. This does not mean everyone adopting Gradualizer has to apply the idiom.

All in all, though, without at least `assert_type` an `integer()` is not an `arity()`, so we need something to convince the type checker.

Reply:
Yeah I know we need `?assert_type` to silence the type error, which we already had in place, but we talked about "inserting runtime casts" as the papers mentioned, which would make the runtime type error happen where the assert is and not somewhere else later on.

With an advanced macro system (Elixir? or code generation?) we could make `?assert_type(Expr, arity())` expand to sth `is_integer(Expr) andalso Expr >= 0 andalso Expr < 256`, but it would require smartness beyond Erlang's macro system and it might not work for all types.

Reply:
#446

Reply:
A few more comments on this:

> we talked about "inserting runtime casts"

I think that if/when we get to actually inserting such casts, we'll have the freedom to drop such explicit checks as the one we're commenting. For now, I think this is the safest we can do, yet still reasonably lightweight.

> With an advanced macro system [...]

Indeed, or if/when running type checking as a stage of a compiler.